### PR TITLE
endpoint: batch get snapshot

### PR DESCRIPTION
**Pull request by overvenus**
It's separated from #2107. Batching get snapshot reduces many callbacks and saves cpu time.

CC #2111

### CODE REVIEW

File: src/coprocessor/endpoint.rs [hunk 1]
```diff
@@ -95,18 +97,68 @@ impl Host {
...
+            }
+        }
+    }
 }
 
 pub enum Task {
     Request(RequestTask),
     SnapRes(u64, engine::Result<Box<Snapshot>>),
+    BatchSnapRes(Vec<(u64, engine::Result<Box<Snapshot>>)>),
+    BacklogRequests(Vec<u64>),
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: Backlog here is still a bad name

Comment (overvenus) [inline src/coprocessor/endpoint.rs]: Yes, I forgot to rename it. How about `Retry` or `RetryRequest`?

File: src/coprocessor/metrics.rs [hunk 2]
```diff
@@ -89,4 +89,12 @@ lazy_static! {
             "Total number of rocksdb query of get or scan count",
             &["type"]
         ).unwrap();
+
+    pub static ref BATCH_REQUEST_TASKS: HistogramVec =
+        register_histogram_vec!(
+            "tikv_coprocessor_batch_request_tasks_total",
+            "Bucketed histogram of total number of a batch request task",
+            &["type"],
+            linear_buckets(0.0, 2.0, 20).unwrap()
```
Comment (siddontang) [inline src/coprocessor/metrics.rs]: change the same with scheduler too

**Comment by siddontang** [comment]
what is the batch size?

**Comment by overvenus** [comment]
Now, it's 256.

File: src/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -256,62 +308,86 @@ impl BatchRunnable<Task> for Host {
...
+                        retry.push(id);
+                    }
+                }
+            }
+
+            if !ready.is_empty() {
+                sched.schedule(Task::BatchSnapRes(ready)).unwrap();
+            }
+            if !retry.is_empty() {
+                sched.schedule(Task::RetryRequests(retry)).unwrap();
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: use BATCH_REQUEST_TASKS with retry label here

File: src/coprocessor/endpoint.rs [hunk 4]
```diff
@@ -256,62 +308,86 @@ impl BatchRunnable<Task> for Host {
...
-                                COPR_PENDING_REQS.with_label_values(&[type_str, pri_str]).dec();
-                            });
-                        } else if pri == CommandPri::High {
-                            self.high_priority_pool.execute(txn_id, move || {
-                                end_point.handle_request(req);
-                                COPR_PENDING_REQS.with_label_values(&[type_str, pri_str]).dec();
-                            });
+                }
+                Task::RetryRequests(retry) => {
+                    BATCH_REQUEST_TASKS.with_label_values(&["retry"])
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: move the statistics down, not proper here, here we don't do batch requests now.

File: src/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -256,62 +308,85 @@ impl BatchRunnable<Task> for Host {
...
                     }
                 }
             }
         }
+
+        if grouped_reqs.is_empty() {
+            return;
+        }
+
+        let mut req_ids = Vec::with_capacity(grouped_reqs.len());
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: seem we don't need req_ids here.
We can ensure that the first id is self.last_req_id + 1, and the last is self.last_req_id + len(grouped_reqs)

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay @zhangjinpeng1987

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Comment by siddontang** [comment]
PTAL @BusyJay @zhangjinpeng1987

**Comment by siddontang** [comment]
@overvenus seem you should rebase to master?

File: src/coprocessor/endpoint.rs [hunk 6]
```diff
@@ -273,69 +333,80 @@ impl BatchRunnable<Task> for Host {
...
+                },
             }
         }
+
+        if grouped_reqs.is_empty() {
+            return;
+        }
+
+        let mut batch = Vec::with_capacity(grouped_reqs.len());
+        let start_id = self.last_req_id + 1;
```
Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: `start_id` is not equal to the first group of reqs' id, is it correct?

Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: My fault, there are equal.

File: src/server/server.rs [hunk 7]
```diff
@@ -30,7 +30,7 @@ use super::resolve::StoreAddrResolver;
 use super::snap::{Runner as SnapHandler, Task as SnapTask};
 use super::raft_client::RaftClient;
 
-const DEFAULT_COPROCESSOR_BATCH: usize = 50;
+const DEFAULT_COPROCESSOR_BATCH: usize = 256;
```
Comment (zhangjinpeng1987) [inline src/server/server.rs]: How much performance has improved by changing batch size from 50 to 256?

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/coprocessor/endpoint.rs [hunk 8]
```diff
@@ -101,18 +103,76 @@ impl Host {
...
+            let pri = req.priority();
+            let pri_str = get_req_pri_str(pri);
+            let type_str = get_req_type_str(req.req.get_tp());
+            COPR_PENDING_REQS
+                .with_label_values(&[type_str, pri_str])
+                .add(1.0);
+            let end_point = TiDbEndPoint::new(snap.clone());
+            let txn_id = req.start_ts.unwrap_or_default();
+
+            if pri == CommandPri::Low {
```
Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: Can we use match here to get the pool to be operated on?

File: src/coprocessor/endpoint.rs [hunk 9]
```diff
@@ -101,18 +103,76 @@ impl Host {
...
+    RetryRequests(Vec<u64>),
 }
 
 impl Display for Task {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         match *self {
             Task::Request(ref req) => write!(f, "{}", req),
             Task::SnapRes(req_id, _) => write!(f, "snapres [{}]", req_id),
+            Task::BatchSnapRes(ref batch) => {
+                let ids: Vec<u64> = batch.iter().map(|&(id, _)| id).collect();
```
Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: This is unnecessary.

File: src/coprocessor/endpoint.rs [hunk 10]
```diff
@@ -101,18 +103,77 @@ impl Host {
...
+            let pri_str = get_req_pri_str(pri);
+            let type_str = get_req_type_str(req.req.get_tp());
+            COPR_PENDING_REQS
+                .with_label_values(&[type_str, pri_str])
+                .add(1.0);
+            let end_point = TiDbEndPoint::new(snap.clone());
+            let txn_id = req.start_ts.unwrap_or_default();
+
+            match pri {
+                CommandPri::Low => {
```
Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: Use match to get pool instead.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**